Persistent Congestion Time Threshold (#2365)
quicwg/base-drafts: draft-ietf-quic-recovery.md

4 * rttvar + max_ack_delay) * ((2 ^ kPersistentCongestionThreshold) - 1).

For example, assume:

  smoothed_rtt = 1
  rttvar = 0
  max_ack_delay = 0
  kPersistentCongestionThreshold = 2

If an eck-eliciting packet is sent at time = 0, the following scenario would
illustrate persistent congestion:

  t=0 | Send Pkt #1 (App Data)
  t=1 | Send Pkt #2 (PTO 1)
  t=3 | Send Pkt #3 (PTO 2)
  t=7 | Send Pkt #4 (PTO 3)
  t=8 | Recv ACK of Pkt #4

The first three packets are determined to be lost when the ACK of packet 4 is
received at t=8.  The congestion period is calculated as the time between the
oldest and newest lost packets: (3 - 0) = 3.  The duration for persistent
congestion is equal to: (1 * ((2 ^ 2) - 1)) = 3.  Because the threshold was
reached, the network is considered to have experienced persistent congestion.

When persistent congestion is established, the sender's congestion window MUST
be reduced to the minimum congestion window (kMinimumWindow).  This response of
collapsing the congestion window on persistent congestion is functionally
similar to a sender's response on a Retransmission Timeout (RTO) in TCP

> {{RFC5681}}.

Reviewer: ```suggestion
{{RFC5681}} after Tail Loss Probes (TLP) {{TLP}}.
```

until an acknowledgement is received that establishes loss or delivery of
packets.

## Persistent Congestion

When an ACK frame is received that establishes loss of all in-flight packets
sent over a long enough period of time, the network is considered to be
experiencing persistent congestion.  Commonly, this can be established by
consecutive PTOs, but since the PTO timer is reset when a new ack-eliciting
packet is sent, an explicit duration must be used to account for those cases
where PTOs do not occur or are substantially delayed.  This duration is the

> equivalent of kPersistentCongestionThreshold consecutive PTOS, and is computed

Reviewer 2: ```suggestion
equivalent of kPersistentCongestionThreshold consecutive PTOs, and is computed
```

as follows:
~~~
(smoothed_rtt + 4 * rttvar + max_ack_delay) *
    ((2 ^ kPersistentCongestionThreshold) - 1)
~~~

For example, assume:

  smoothed_rtt = 1
  rttvar = 0
  max_ack_delay = 0
  kPersistentCongestionThreshold = 2

If an eck-eliciting packet is sent at time = 0, the following scenario would
illustrate persistent congestion:

  t=0 | Send Pkt #1 (App Data)
  t=1 | Send Pkt #2 (PTO 1)
  t=3 | Send Pkt #3 (PTO 2)
  t=7 | Send Pkt #4 (PTO 3)
  t=8 | Recv ACK of Pkt #4

The first three packets are determined to be lost when the ACK of packet 4 is
received at t=8.  The congestion period is calculated as the time between the

> oldest and newest lost packets: (3 - 0) = 3.  The duration for persistent

Reviewer 3: This doesn't match the description in the first paragraph, does it?
> When an ACK frame is received that establishes loss of all in-flight packets, sent over a long enough period of time, the network is considered to be experiencing persistent congestion.

With this algorithm, we would detect persistent congestion even if packet 2 and 4 were acked, right?

Author: The text says `all in-flight packets` meaning a contiguous block of sent packets. The example doesn't call out that check for a contiguous block explicitly, but it would have to be made. So, in that case if packets 2 and 4 were acknowledged you wouldn't have that contiguous block of time that meets the requirements.

Reviewer 2: I think this example works, but the pseudocode below is not quite right for this reason.  The easiest way to fix it would be to assume no packet number gaps and check if the number of packets lost is equal to largest - smallest + 1.

Author: Do we need to really spell this out with a 100% complete pseudocode example? I can imagine a couple ways for implementations to achieve this and I don't think we need to call them all out.

Reviewer 2: The text is normative and the pseudocode is an appendix.  I'm ok with the pseudocode not being perfect, especially for cases that are fairly unlikely in practice.

Reviewer 3: The point of the pseudo-code is not to tell implementations exactly what to do. You can still implement it any way you want.
I'd like to avoid inconsistencies between the text and the pseudo code. We had these before when we introduced packet number spaces, and it now turns out to be quite a lot of work to get things straight.

> congestion is equal to: (1 * ((2 ^ 2) - 1)) = 3.  Because the threshold was

Author: We pretty clearly state the equation for calculating the persistent congestion window above and state that `kPersistentCongestionThreshold` is assumed to be 2 for this example. I don't really want to make this section any more verbose.

> congestion is equal to: (1 * ((2 ^ 2) - 1)) = 3.  Because the threshold was

Reviewer 3: Just saying, this was my experience when reading this paragraph for the first time. Would be interesting to hear what other people reading this text for the first time think.

> congestion is equal to: (1 * ((2 ^ 2) - 1)) = 3.  Because the threshold was

Author: Do you have a recommendation of some specific text to add or change that would make things more clear?

> congestion is equal to: (1 * ((2 ^ 2) - 1)) = 3.  Because the threshold was

Reviewer 3: `1 * ((2 ^ kPersistentCongestionThreshold) - 1)) = 3`

experiencing persistent congestion.  Commonly, this can be established by
consecutive PTOs, but since the PTO timer is reset when a new ack-eliciting
packet is sent, an explicit duration must be used to account for those cases
where PTOs do not occur or are substantially delayed.  This duration is the
equivalent of kPersistentCongestionThreshold consecutive PTOs, and is computed
as follows:
~~~
(smoothed_rtt + 4 * rttvar + max_ack_delay) *
    ((2 ^ kPersistentCongestionThreshold) - 1)
~~~

For example, assume:

  smoothed_rtt = 1
  rttvar = 0
  max_ack_delay = 0
  kPersistentCongestionThreshold = 2

If an eck-eliciting packet is sent at time = 0, the following scenario would
illustrate persistent congestion:

  t=0 | Send Pkt #1 (App Data)
  t=1 | Send Pkt #2 (PTO 1)
  t=3 | Send Pkt #3 (PTO 2)
  t=7 | Send Pkt #4 (PTO 3)
  t=8 | Recv ACK of Pkt #4

The first three packets are determined to be lost when the ACK of packet 4 is
received at t=8.  The congestion period is calculated as the time between the
oldest and newest lost packets: (3 - 0) = 3.  The duration for persistent
congestion is equal to: (1 * ((2 ^ kPersistentCongestionThreshold) - 1)) = 3.

> Because the threshold was reached, the network is considered to have experienced

Reviewer: Suggestion: "Because the threshold was reached and because none of the packets between the oldest and the newest packets are acknowledged, the network is considered ..."

until an acknowledgement is received that establishes loss or delivery of
packets.

## Persistent Congestion

When an ACK frame is received that establishes loss of all in-flight packets
sent over a long enough period of time, the network is considered to be
experiencing persistent congestion.  Commonly, this can be established by

> consecutive PTOs (pto_count is more than kPersistentCongestionThreshold, see

Reviewer 2: ```suggestion
consecutive PTOs
```

> consecutive PTOs (pto_count is more than  kPersistentCongestionThreshold, see

Reviewer 2: ```suggestion
consecutive PTOs (pto_count is more than kPersistentCongestionThreshold, see
```

Reviewer 2: Also, I'd remove the () here and then define what the actual period is at the end of this paragraph.

{{cc-consts-of-interest}}), but since the PTO timer is reset when a new

> ack-eliciting packet is sent, an explicit timeout must be used to account for

Reviewer 2: I don't think it's really a timeout.  How about period or duration or time?

Author: I think time or duration might fit.

> those cases where PTOs do not occur.

Reviewer 2: ```suggestion
those cases where PTOs do not occur or are substantially delayed.
```

until an acknowledgement is received that establishes loss or delivery of
packets.

## Persistent Congestion

When an ACK frame is received that establishes loss of all in-flight packets
sent over a long enough period of time, the network is considered to be
experiencing persistent congestion.  Commonly, this can be established by
consecutive PTOs, but since the PTO timer is reset when a new ack-eliciting
packet is sent, an explicit duration must be used to account for those cases
where PTOs do not occur or are substantially delayed.  This duration is the

> equivalent of kPersistentCongestionThreshold consecutive PTOS, smoothed_rtt +

Reviewer 2: ```suggestion
equivalent of kPersistentCongestionThreshold consecutive PTOS: smoothed_rtt +
```

> 4 * rttvar + max_ack_delay * ((2 ^ kPersistentCongestionThreshold) - 1).

Reviewer 2: Q for here and below: Should max_ack_delay be added after exponential backoff, instead of before?  Adding it after incentivizes implementations to indicate a realistic value, since an unrealistic one just adds to rttvar, which seems desirable.

Author: The order is the same in `SetLossDetectionTimer`:
```    // Calculate PTO duration
    timeout =
      smoothed_rtt + 4 * rttvar + max_ack_delay
    timeout = max(timeout, kGranularity)
    timeout = timeout * (2 ^ pto_count)```
Though, on a side issue, I feel the `kGranularity` check should be moved to the end of that calculation...

Reviewer 2: Agreed, what you have matches the current pseudocode.  I'll open an issue for kGranularity and max_ack_delay and when they should be added.

Reviewer: Agreed that it seems odd to backoff on max_ack_delay. @ianswett : Let's do that separately.

until an acknowledgement is received that establishes loss or delivery of
packets.

> ## Persistent Congestion

Reviewer 2: I think this should be a sub-section of Probe Timeout?

Author: I'm not super opinionated either way, but I feel like coupling the two things go us into this problem in the first place. They are related but still fairly independent in my mind. Especially, if we ever go with the model of a separate timer.

The network is considered to be experiencing persistent congestion when the
sender goes a period of time without receiving any ACK frames for its in-flight,

> ack-eliciting packets.  Generally, this can be calculated as a number of

Reviewer 4: I'm getting used to that word ...

consecutive PTOs (pto_count is more than  kPersistentCongestionThreshold, see

> {{cc-consts-of-interest}}), but since a PTO only occurs if no other

Reviewer 2: How about "but since the PTO timer is reset when a new ack-eliciting packet is sent,"

Author: Sure.

> ack-eliciting packets have been sent, an explicit time out must be used to

Author: Could you make these suggestions on the latest changes so I can accept them more easily?

until an acknowledgement is received that establishes loss or delivery of
packets.

## Persistent Congestion

> The network is considered to be experiencing persistent congestion when the

Reviewer 2: The old text was more correct than this text in the sense that it's important that not only does time pass, but all the packets are lost within that period of time.

How about:
"When an ACK frame is received that establishes loss of all in-flight packets sent over a long enough period of time, the network is considered to be experiencing persistent congestion."

Author: For your suggestion, do you really need the word "all" in there. To me, the loss of any in-flight packets over a long enough time seems to be the requirement (at least how we have the pseudocode). It think the removal the word "all" might be better. Other than that, I am fine with making the change.

Reviewer 2: It has to be all packets lost within that period of time(ie: total blackout).  And now that you mention it, that means the pseudocode isn't quite right either.

As #2255 details, the goal is to be consistent with the previous behavior in terms of when persistent congestion would be established.

Author: I'm not sure how we'd update the pseudocode to indicate loss of all packets. I'd have to think a bit more about it.

sender goes a period of time without receiving any ACK frames for its in-flight,

> ack-eliciting packets.  Generally, this can be calculated as a number of

Reviewer 2: Suggestion: replace "Generally, this can be calculated as a number of consecutive PTOs (),..." with
"Commonly, this can be established by consecutive PTOs,..."

consecutive PTOs (pto_count is more than  kPersistentCongestionThreshold, see
{{cc-consts-of-interest}}), but since the PTO timer is reset when a new

> ack-eliciting packet is sent, an explicit time out must be used to account for

Reviewer 4: ```suggestion
ack-eliciting packet is sent, an explicit timeout must be used to account for
```

> those cases where PTOs don't occur.

Reviewer 4: ```suggestion
those cases where PTOs do not occur.
```

> When persistent congestion is encountered the sender's congestion window MUST be

Reviewer 2: ```suggestion
When persistent congestion is established, the sender's congestion window MUST be
```

~~~
   congestion_window = kInitialWindow

>    bytes_in_flight = 0

Reviewer 4: `probe_pending = 0`

~~~
   OnPacketSentCC(bytes_sent):

>      if (bytes_in_flight == 0):

Author: That doesn't quite work. What if the first packet isn't ack_eliciting but the second is? The first would still increment the counter, so when the second comes in it wouldn't be 0 any more.

>      if (bytes_in_flight == 0):

Reviewer 4: right

>      if (bytes_in_flight == 0):

Reviewer 4: ```
if (ack_eliciting && (bytes_in_flight == 0 || probe_pending)):
    probe_pending = 0
    probe_start_time = Now()
else:
   probe_pending = 1

>      if (bytes_in_flight == 0):

Reviewer 4: It's too late, I keep getting this wrong - wait

>      if (bytes_in_flight == 0):

Reviewer 4: Edited above comment, I think it is ok now

>        probe_start_time = Now()

Reviewer 2: It's important not to start this if the packet isn't ack-eliciting, but I will say that requirement makes this more complex, since you have to track the first ack-eliciting packet to be sent.

Author: So, are you suggestion a change? Perhaps a comment?

   OnPacketAckedCC(acked_packet):
     // Remove from bytes_in_flight.
     bytes_in_flight -= acked_packet.size
     if (bytes_in_flight != 0):

>        probe_start_time = Now()

Reviewer 2: I don't think it's necessary to check bytes_in_flight != 0 here?

Author: Yeah, you are right. I'm not sure what I was thinking when I added this. I will remove it.

are detected lost.

~~~
   InPersistentCongestion(oldest_loss_time):
     elapsed_time = Now() - oldest_loss_time
     pto = smoothed_rtt + 4 * rttvar + max_ack_delay
     return elapsed_time > pto * (2 ^ kPersistentCongestionThreshold - 1)

   OnPacketsLost(lost_packets):
     // Remove lost packets from bytes_in_flight.
     for (lost_packet : lost_packets):
       bytes_in_flight -= lost_packet.size
     smallest_lost_packet = lost_packets.first()

>      largest_lost_packet = lost_packets.last()

Author: Slight tangent to this PR: what do folks think about changing these variables to `oldest_lost_packet` and `newest_lost_packet`?

Reviewer 2: I don't have a strong preference either way.

are detected lost.

~~~
   InPersistentCongestion(oldest_loss_time):

>      return

Reviewer 2: This expression is hard to read, I think it's worth breaking it up some.
 elapsed_time = Now() - oldest_loss_time
 pto = smoothed_rtt + 4 * rttvar + max_ack_delay
 return elapsed_time > pto * (2 ^ kPersistentCongestionThreshold - 1)

are detected lost.

~~~
   InPersistentCongestion(oldest_loss_time):

>      elapsed_time = Now() - oldest_loss_time

Reviewer 2: ```suggestion
     congestion_period = Now() - oldest_loss_time
```

are detected lost.

~~~
   InPersistentCongestion(oldest_loss_time):
     congestion_period = Now() - oldest_loss_time
     pto = smoothed_rtt + 4 * rttvar + max_ack_delay
     return
       congestion_period >
         pto * (2 ^ kPersistentCongestionThreshold - 1)

   OnPacketsLost(lost_packets):
     // Remove lost packets from bytes_in_flight.
     for (lost_packet : lost_packets):
       bytes_in_flight -= lost_packet.size
     oldest_lost_packet = lost_packets.first()
     newest_lost_packet = lost_packets.last()

     // Start a new congestion epoch if the last lost packet
     // is past the end of the previous recovery epoch.
     CongestionEvent(newest_lost_packet.time_sent)

     // Collapse congestion window if persistent congestion

>      if (InPersistentCongestion(oldest_lost_packet.time_sent)):

Reviewer 2: I'm starting to think we should pass in congestion_period and make it oldest_lost_packet - newest_lost_packet.

I think using Now() may be incorrect.  What if one packet is lost and then a few ACKs are lost.  In this case, Now() is substantially larger than oldest_lost_packet, but in fact only one packet was lost.

Reviewer 2: I'd like the opinion of @janaiyengar here.  I think this is likely fairly good, but I think passing in newest_lost_time is easy and more correct, given they're easily available in the pseudocode.

Author: Let me propose an example. Assume smooth_rtted = 1 sec, rttvar = 0, max_ack_delay = 0 for simplicity's sake.
```
t=0 | Send Pkt #1 (App Data)
t=1 | Send Pkt #2 (PTO 1)
t=3 | Send Pkt #3 (PTO 2)
t=7 | Send Pkt #4 (PTO 3)
t=8 | Recv ACK of Pkt #4
```
At t=8 you'd get a congestion event for packets 1 & 2 being lost. If you use Now() for the persistent congestion calculation, this would be declared as persistent congestion. If you used newest_lost_time it wouldn't. Which behavior do we want?

Reviewer 2: If 4 is acked and the others are lost, then it should be persistent congestion.

But if 2, 3 and 4 were all acked at once(t=8), that should not be persistent congestion.

Author: Hmm, well then we'd have to make two changes to achieve that behavior.
1. Use `newest_lost_packet - oldest_lost_packet` for the time window.
2. Make the threshold 2 PTO instead of 3.

Without the second part above, an ACK of pkt 4 would produce a loss window of only 3 seconds (3 - 0), right?

Reviewer 2: That sounds correct to me.  Can you add your example as an example to the text section as well?  This is clearly very confusing, so an example would help.

Author: Sure. Do you want me to describe the example in just words, or put almost exactly what I have above with the text?

Reviewer 2: I think how you have it laid out is great and an effort to convert it into text will probably make it less clear.  But I'm not familiar enough with markdown to know how to get the formatting right.

Author: Ok. I'll give it a go.

are detected lost.

~~~
   InPersistentCongestion(oldest_loss_time):

>      congestion_period = Now() - oldest_loss_time

Reviewer 2: I think I concluded newest_lost_time - oldest_loss_time is the best option here.  Using Now() means that a single lost packet combined with some temporary ack loss causes this to trigger, which I do not believe is the intent.

Reviewer: Yes, I agree. I had thought that Now() was best, but that does not account for pauses in sending... newest_lost_time - oldest_lost_time is quite good I think.

are detected lost.

~~~
   InPersistentCongestion(congestion_period):
     pto = smoothed_rtt + 4 * rttvar + max_ack_delay
     return
       congestion_period >

>          pto * (2 ^ kPersistentCongestionThreshold - 1)

Reviewer: editorial nit: this looks a bit confusing because 'return' is on a line by itself. How about:

return congestion_period >
                    pto * (2 ^ kPersistentCongestionThreshold - 1)

are detected lost.

~~~

>    InPersistentCongestion(congestion_period):

Reviewer: Suggested rework of this function:
```
InPersistentCongestion(newest_lost_packet):
    pto = smoothed_rtt + 4 * rttvar + max_ack_delay
    congestion_period = pto * (2 ^ kPersistentCongestionThreshold - 1)
    // Find the latest packet that was sent at least pto time before the newest lost packet
    oldest_sent_packet = FindPacketSentBefore(newest_lost_packet, congestion_period)
    // Determine if all packets in the window, including the  edges, are marked lost
    return IsWindowLost(oldest_sent_packet, newest_lost_packet)
```

Author: I feel like this overly complicates things (at least for the pseudocode). What if simplify it a bit, like the following:
```
InPersistentCongestion(newest_lost_packet):
    pto = smoothed_rtt + 4 * rttvar + max_ack_delay
    congestion_period = pto * (2 ^ kPersistentCongestionThreshold - 1)
    // Determine if all packets in the window before the newest lost packet,
    // including the edges, are marked lost
    return IsWindowLost(newest_lost_packet, congestion_period)
```

     pto = smoothed_rtt + 4 * rttvar + max_ack_delay
     return congestion_period >
       pto * (2 ^ kPersistentCongestionThreshold - 1)

   OnPacketsLost(lost_packets):
     // Remove lost packets from bytes_in_flight.
     for (lost_packet : lost_packets):
       bytes_in_flight -= lost_packet.size
     oldest_lost_packet = lost_packets.first()
     newest_lost_packet = lost_packets.last()

     // Start a new congestion epoch if the last lost packet
     // is past the end of the previous recovery epoch.
     CongestionEvent(newest_lost_packet.time_sent)

     // Collapse congestion window if persistent congestion

>      if (InPersistentCongestion(

Reviewer: Change params to InPersistentCongestion to only have newest_lost_packet.

are detected lost.

~~~
   InPersistentCongestion(newest_lost_packet):
     pto = smoothed_rtt + 4 * rttvar + max_ack_delay
     congestion_period =
       pto * (2 ^ kPersistentCongestionThreshold - 1)
     // Determine if all packets in the window before the
     // newest lost packet, including the edges, are marked
     // lost
     return IsWindowLost(newest_lost_packet, congestion_period)

   OnPacketsLost(lost_packets):
     // Remove lost packets from bytes_in_flight.
     for (lost_packet : lost_packets):
       bytes_in_flight -= lost_packet.size

>      newest_lost_packet = lost_packets.last()

Reviewer 2: Elsewhere we use largest(ie: largest_acked) and there are currently no uses of newest in the recovery draft, so I'm going to change this back to largest.  If we want to start using newest, let's do that in a separate PR.

Author: Fine with me. I might look at changing it later.
